sycl/doc/design/OptionalDeviceFeatures.md

```diff
@@ -913,211 +913,253 @@ aspects in the array.
-
-##### AOT target identification
-
-There are several user-visible places in the SDK where SYCL device target
-architectures need to be identified:
-
-- `-fsycl-targets` option
-- a device configuration file entry
-- `-target` option of the `sycl-aspect-filter` tool
-- a new SYCL enumeration named `architecture`
-
-The following table lists these target names:
-
-| name                 | has `architecture` enum | description                            |
------------------------|-------------------------|----------------------------------------|
-| ptx64                | no                      | Generic 64-bit PTX target architecture |
-| spir64               | no                      | Generic 64-bit SPIR-V target           |
-| x86\_64              | yes                     | Generic 64-bit x86 architecture        |
-| intel\_gpu\_\<name\> | yes                     | Intel graphics architecture \<name\>   |
-
-The "name" column in this table lists the possible target names.  Since not all
-targets have a corresponding enumerator in the `architecture` enumeration, the
-second column tells when there is such an enumerator.  The last row in this
-table corresponds to all of the architecture names listed in the
-[sycl\_ext\_oneapi\_device\_architecture][8] extension whose name starts with
-`intel_gpu_`.
-
-[8]: <../extensions/experimental/sycl_ext_oneapi_device_architecture.asciidoc>
+After the header line, there is one line for each IR module that the
+`sycl-post-link` tool generates.
 
-TODO: This table needs to be filled out for the CPU variants supported by the
-`opencl-aot` tool (avx512, avx2, avx, sse4.2) and for the FPGA targets.  We
-also need to figure out how CUDA fits in here.
+To implement the final filtering pass, the `sycl-post-link` tool uses the
+configuration file to determine the features (e.g. aspects or sub-group sizes)
+that are supported by a target device.  If any IR module uses an incompatible
+feature, the corresponding row in the table is changed to use an empty IR
+module instead.  To illustrate, let's assume that
+`-o=intel_gpu_pvc,output-pvc.table` is passed and that module "mod2.bc" uses
+features that are incompatible with PVC.  The generated table file
+"output-pvc.table" would look like this:
 
-Example of clang compilation invocation with 2 AOT targets and generic SPIR-V:
 ```
-clang++ -fsycl -fsycl-targets=spir64,intel_gpu_12_0_0,ptx64 ...
+[Code|Properties|Symbols]
+mod1.bc|mod1.prop|mod1.sym
+empty.bc|mod2.prop|mod2.sym
+mod3.bc|mod3.prop|mod3.sym
 ```
 
+Notice how the incompatible IR module "mod2.bc" has been replaced with an empty
+IR module "empty.bc".  The other entries on this row ("mod2.prop" and
+"mod2.sym") are unchanged.  It's important to preserve the original properties,
+so that the runtime can diagnose an error if the application mistakenly tries
+to submit a kernel from this IR module to a device that does not support the
+features it uses.
```

Comment:
> If runtime can't find an image which is compatible with a device then corresponding error will be thrown and it should be enough indication for a user that this kernel can't be used on that device and we can adjust our exception messages for that.

The runtime uses the "SYCL/device requirements" property set to construct a meaningful error message like:

```
Kernel uses optional feature corresponding to 'aspect::fp16' but device does
not support this aspect.
```

If we delete the row with the empty device image, we also delete the property set, so it becomes impossible to raise a helpful error message like this.

> Moreover, if there is no PVC-specific version of a kernel, but there is a generic SPIR-V version of a kernel, then we can silently fallback to JIT and everything will work just fine. As a user, that is exactly what I would expect from the implementation.

This won't work because the SPIR-V module uses some feature that is not supported on PVC in this case.  (That's the only reason we substitute the empty device image.)  Therefore, we must ensure that the runtime does not try to JIT compile the SPIR-V in this case.  If it did invoke the JIT compiler, it would fail with an exception.

> less temporary files, less I/O, less work for linker, less work at runtime to skip through those empty device images, etc

This addresses only part of your concern, but `sycl-post-link` would only need to generate a single "empty.bc" file, and that file can be shared by all rows in all tables that need an empty entry.  Also, the runtime does not need to skip over the empty entries.  In fact, we want the runtime to find the empty entry, so it can find the corresponding property set.